### PR TITLE
Add [CurseForge] badges

### PR DESCRIPTION
**Pull request by SandroHc**
Adds the following badges:
| URL | Example | Result |
| --- | --- | --- |
| /curseforge/dt/:projectId | /curseforge/dt/238222 | Downloads - /curseforge/dt/238222 |
| /curseforge/game-versions/:projectId | /curseforge/game-versions/238222 | Game versions - /curseforge/game-versions/238222 |
| /curseforge/v/:projectId | /curseforge/v/238222 | Latest version - /curseforge/v/238222 |

To generate the secret **CURSEFORGE_API_KEY** (yml: private.curseforge_api_key) please [signup to the CurseForge Console][cf signup] with a Google account and create an organization, then go to the [API keys page][cf api key] and copy the generated API key. I can provide the API key I used for my own testing if creating a new one is too much of an hassle.

~~Also added the following logos:~~ (Removed, as the logos were already available in the [Simple Icons library](https://simpleicons.org/?q=curseforge))

| Logo | Location | Source |
| ------------- | ------------- | --- |
| CurseForge logo | ~~/logo/curseforge.svg~~ | ~~https://static-beta.curseforge.com/images/favicon.ico~~ |
| Modrinth logo | ~~/logo/modrinth.svg~~ | ~~https://modrinth.com/favicon.ico~~ |

[cf api]: https://docs.curseforge.com
[cf signup]: https://console.curseforge.com/#/signup
[cf api key]: https://console.curseforge.com/#/api-keys

### CODE REVIEW

**Comment by calebcartwright** [comment]
Thank you for the PR, although functionally I feel like this duplicates #7499

Are there any material differences with this PR that you think should be considered over #7499, and/or do you think anything has changed or that we should be aware of wrt the signup process? The seeming complexities around that were largely what things have been blocked on for a while

**Comment by SandroHc** [comment]
Hey, thanks for the reply!

Regarding #7499, the only difference are two extra endpoints ("v/:projectId" and "game-versions/:projectId") and the tests, which are trivial to implement once the API token issue gets sorted out. I am okay with closing this PR in favour of #7499; the extra features can be introduced later. Should have checked if there was any previous work done beforehand, so my fault.

**Comment by PyvesB** [comment]
We have a Google account for Shields.io that we could reuse here. Alternatively I've got a Forge account I'm be happy to share for this matter. https://github.com/badges/shields/pull/7499 links to a form, happy to fill that out and submit it, it doesn't look nearly as annoying than the Youtube form I had to deal with yesterday (#8605). :)

**Comment by SandroHc** [comment]
@PyvesB, that form is actually outdated! The new way is much simpler and just requires signing up with a Google account an creating an organization; no manual review needed. Please see the description of this PR for instructions. :)

The only question is what is the request cap per token, as [the documentation](https://docs.curseforge.com) doesn't talk about rate limits.

**Comment by PyvesB** [comment]
> @PyvesB, that form is actually outdated! The new way is much simpler and just requires signing up with a Google account an creating an organization; no manual review needed. Please see the description of this PR for instructions. :)

Thanks, it indeed is easy to get an API key. Though it's a bit unfortunate that you can only have one key per org, we'll potentially have to use someone's personal account for CI.

> The only question is what is the request cap per token, as [the documentation](https://docs.curseforge.com/) doesn't talk about rate limits.

Do you think you could drop the Curse Forge team a line to figure this out? Hitting various API rate limits has been a recurring problem for Shields.io, we wouldn't want to introduce a new badge if a few months down the line badges are breaking for users due to an unforeseen limit being hit. :)

**Comment by SandroHc** [comment]
Sent an email (as per [this page](https://support.curseforge.com/en/support/solutions/articles/9000205544-contact-us)) asking about the rate limits. Will give an update once I get a reply.

**Comment by SandroHc** [comment]
@PyvesB, they replied back asking for the expected traffic. Could I add you to the conversation, so that you could answer those hard-hitting questions that I don't have the answers to?

**Comment by calebcartwright** [comment]
> @PyvesB, they replied back asking for the expected traffic. Could I add you to the conversation, so that you could answer those hard-hitting questions that I don't have the answers to?

Short answer is yes to plugging the Shields team into the thread, but will also say that it's rather hard for us to provide anything other than an uninformed guestimate with something like this.

With a platform used as widely and heavily like GitHub, with as large a user base, we can generate a lot of traffic in the delivery of badges (we peak somewhere in the 4k-5k requests/minute with GitHub). We've got other badges where we only make a handful of API requests per day (e.g. less than 100).

It's ultimately a factor of the number of users of the platform, and the subset of those that want badges displaying data from/about that platform

**Comment by PyvesB** [comment]
For anyone following this PR: we are currently talking to the Curse Forge team. Please bear with us whilst we exchange information with them. :)

**Review by PyvesB — COMMENTED** [review]
The CurseForge team has responded:
> Please move forward with the integration as is, and if anything should come up - please reach out to us again.

As the immediate next step, I've reviewed the code. Thanks for the contribution @SandroHc, this looks like great work to me! A few small changes requested inline. :)

One issue is that I've not managed to add an API key for GitHub Actions, I get the following error:
````
Failed to add secret. Secret names can only contain alphanumeric characters ([a-z], [A-Z], [0-9]) or underscores (_). Spaces are not allowed. Must start with a letter ([a-z], [A-Z]) or underscores (_). 
````

Indeed, their secret contains character not listed above. Any ideas @chris48s ?

File: config/local.template.yml
Comment (SandroHc) [reply config/local.template.yml]: Slipped through when I was cleaning up the PR. I've removed it.

File: services/modrinth/modrinth-downloads.service.js
Comment (SandroHc) [reply services/modrinth/modrinth-downloads.service.js]: Done

File: services/curseforge/curseforge-downloads.service.js
Comment (SandroHc) [reply services/curseforge/curseforge-downloads.service.js]: Done

**Comment by SandroHc** [comment]
I am not sure what the lint CI step [is complaining about](https://github.com/badges/shields/actions/runs/5675050879/job/15379683071?pr=9252). Both files are fine when I check them locally:
```bash
$ npx prettier --check "core/base-service/auth-helper.js"       
Checking formatting...
All matched files use Prettier code style!

$ npx prettier --check "services/curseforge/curseforge-base.js"
Checking formatting...
All matched files use Prettier code style!
```

**Comment by PyvesB** [comment]
We recently updated to a new major version of Prettier (#9357). Could you make sure you're based off of a recent-enough `master` revision, and that you've installed the latest Prettier (`npm ci`)?

**Comment by SandroHc** [comment]
That fixed it. All checks are passing now!

File: services/curseforge/curseforge-base.js
Comment (chris48s) [inline services/curseforge/curseforge-base.js:43]:
```suggestion
        httpErrors: {
```

We changed the name of this param in https://github.com/badges/shields/pull/9233

**Comment by PyvesB** [comment]
In the meantime we've been setting up the tokens on our servers, so whilst we've been a bit slow on the review cycles since this was first opened, this _should_ be fully operational as soon as we merge and kick off a deploy. Thanks for the patience! 👍🏻

File: services/curseforge/curseforge-base.js
Comment (SandroHc) [reply services/curseforge/curseforge-base.js]: Change done and confirmed that it works

image

**Comment by SandroHc** [comment]
All changes are done now! I have retested all three badges to make sure nothing broke in the few months since I opened this PR and everything is looking good.

**Review by PyvesB — APPROVED** [review]

File: services/curseforge/curseforge-downloads.service.js
Comment (PyvesB) [reply services/curseforge/curseforge-downloads.service.js]: There were a few leftovers, I went ahead and removed them myself to avoid another review cycle :)

**Comment by PyvesB** [comment]
I've deployed this PR to staging and production. The new badges are happily working, for example:
 

Thanks again for working with us on this @SandroHc!

**Pull request merged**
